### DISCUSSION TITLE
[DOC] - Missing documentation on how to external repo via registries.yaml

### DISCUSSION THREAD

**Discussion by alex-ioma** [rancher/rke2#6096] [Q&A] [answered]
**Environmental Info:**
RKE2 Version: 1.28.10+rke2r1

Node(s) CPU architecture, OS, and Version: N/A

Cluster Configuration: 3 master + 3 workers

**Describe the bug:**

Apologies for opening this ticket but I could not find a way to add GCP Artifact Registry to the RKE2 containerd runtime via the `registries.yaml` file. 

**Steps To Reproduce:**

- Create a Service Account Key on GCP/IAM with access to GCP Artifact Registry
- Test registry access with said key (it correctly pull from local)
- Try and add key but cannot figure out which how to complete the `registries.yaml` config file.

**Expected behavior:**

Be able to pull images via `crictl` or inside k8s.

**Actual behavior:**

Pull doesn't work - failing with: 
`Failed to pull image "europe-docker.pkg.dev/cm-controls/lgm/lgm-hmi:v0.1.0-alpha": failed to pull and unpack image "europe-docker.pkg.dev/cm-controls/lgm/lgm-hmi:v0.1.0-alpha": failed to resolve reference "europe-docker.pkg.dev/cm-controls/lgm/lgm-hmi:v0.1.0-alpha": failed to authorize: failed to fetch anonymous token: unexpected status from GET request to https://europe-docker.pkg.dev/v2/token?scope=repository%3Acm-controls%2Flgm%2Flgm-hmi%3Apull&service=europe-docker.pkg.dev: 403 Forbidden
`
**Additional context / logs:**

It's likely an error on my end on how to properly configure `registries.yaml` but I feel the documentation can be improved. Also it is not clear if this configuration should be applied to all nodes or only to one master and/or worker and is automatically replicated. 

Thank you in advance for the support.

**Comment by brandond** [comment]
It looks like you're using `europe-docker.pkg.dev` as the registry, so you'd want to add an entry to the `config` section with credentials for `europe-docker.pkg.dev`. This is covered at https://docs.rke2.io/install/containerd_registry_configuration#configs

Registries.yaml is not a cluster resource. The configuration is local to each node, so you must place this configuration on every node that you want to be able to authenticate to your registry. This is also covered at the top of the above-linked page.

**Reply by alex-ioma** [reply]
Thank you @brandond for clarifying the local-only behavior. However, what is documented at the linked page is what I followed without success. 

The issues arise from the fact that Google Artifact Registry uses Service Accounts Key, which are in the form: 
```json
{
  "type": "service_account",
  "project_id": "cm-controls",
  "private_key_id": "random-key-id",
  "private_key": "-----BEGIN PRIVATE KEY-----\nREDACTED\n-----END PRIVATE KEY-----\n",
  "client_email": "registry-ro@cm-controls.iam.gserviceaccount.com",
  "client_id": "106584479505621107395",
  "auth_uri": "https://accounts.google.com/o/oauth2/auth",
  "token_uri": "https://oauth2.googleapis.com/token",
  "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
  "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/user@project.iam.gserviceaccount.com",
  "universe_domain": "googleapis.com"
}
``` 

Whereas RKE2 docs (as well as K3S) only reference an example that is based on simple auth (user/pwd). In specific, these entries should be based on: 
- Service Account Key-based or token-based authentication
- Registry exposed via HTTPS with server-side remote secret (typically Let'sEncrypt) (it is not clear if :443 can be omitted in the `host:port` mirror section)

As mentioned, I've successfully achieved the authentication and image pull by adding name-spaced .dockerconfig manifest with similar contet: 
```json
{"auths":{"europe-docker.pkg.dev":{"auth":"base64-encoded-service-account-key"}}}
```
But could not get it working with the `registry.yaml` approach.

The goal would be to use the latter as it provides access to registry to all namespaces with a single config file that can be easily distributed to the nodes, other than using the approached based on name-spaced secrets. 

The same issues apply while trying to setup Gitlab registry via deploy token. There is not a well-documented way on how to format/pass token and/or user/tokenId into the yaml file and RKE2 is not able to pull from the repo (same erro as per original post). 

In my opinion, the documentation strongly needs improvement, expanding on the above example and clearly stating the limitation of such config. Also I'm yet to find a working configuration for the registries.yaml file. Also, I believe it would be really beneficial if such setting is immediately distributed cluster-wide after applying it to the first note - perhaps via etcd?

I look forward to understand more on how to work with `registries.yaml` file and thank you for support.

**Reply by brandond** [reply]
> it is not clear if :443 can be omitted in the host:port mirror section

If you're pulling the image as `example.com:443/foo/bar:latest` then you would need to specify it as `example.com:443`. If you're not including the port in the image spec when pulling it, then you do not. You just need to be consistent with how you refer to the registry.

> But could not get it working with the registry.yaml approach.

Isn't that covered by this portion of the docs?
>> The credentials consist of either username/password or authentication token:
>> username: user name of the private registry basic auth
>> password: user password of the private registry basic auth
>> auth: authentication token of the private registry basic auth

```yaml
configs:
  europe-docker.pkg.dev:
    auth:
      auth: base64-encoded-service-account-key
```

I think that should do what you want? We don't really cover that in the docs because as far as I know the auth tokens expire fairly quickly so people don't usually bother embedding them in their config.

**Reply by alex-ioma** [reply]
Thank you @brandond 

It depends: indeed token on GCP are short-lived (60 mins) access token to pull from registry - but they might also last longer in case of GitLab Deploy Token. 

In the specific case of GCP a service account key is an RSA key that is used to get a JWT bearer token with which docker / containers should generate an access token. 

Hence it is not clear if: 
1. This key is supported by rke2 containerd implementation 
2. How the key (which is a json- formatted file) should be included in the yaml manifest. Should it be  a base64 encoded? (You mentioned it in your last thread but it is not mentioned in the docs. 

Thank you.

**Reply by brandond** [reply]
Fields in `registries.yaml` are just passed through into the containerd configuration, so it's not really a question of rke2 supporting it or not. The `auth` section is passed through into the same section in the containerd CRI config structure, which is documented upstream here: https://github.com/containerd/containerd/blob/release/1.7/docs/cri/registry.md

Note that you'll need to expand the `Show the original content (DEPRECATED)` docs and find the `Configure Registry Credentials` section. CRI registry auth is technically deprecated by containerd but is not slated for removal, as there is no direct replacement for it.

> To configure a credential for a specific registry, create/modify the /etc/containerd/config.toml as follows:
> ```toml
> # explicitly use v2 config format
> version = 2
> 
> # The registry host has to be a domain name or IP. Port number is also
> # needed if the default HTTPS or HTTP port is not used.
> [plugins."io.containerd.grpc.v1.cri".registry.configs."gcr.io".auth]
>   username = ""
>   password = ""
>   auth = ""
>   identitytoken = ""
> ```
>  The meaning of each field is the same with the corresponding field in .docker/config.json.

**Reply by alex-ioma** [reply]
By only using the auth field in registries.yaml might actually work. However, I'm not able to validate as I get te following errors: 

```bash
crictl pull europe-docker.pgk.dev/some-path/image:v0.1.0-alpha

WARN[0000] image connect using default endpoints: [unix:///var/run/dockershim.sock unix:///run/containerd/containerd.sock unix:///run/crio/crio.sock unix:///var/run/cri-dockerd.sock]. As the default settings are now deprecated, you should set the endpoint instead.
E0604 23:20:24.104891  288825 remote_image.go:171] "PullImage from image service failed" err="rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial unix /var/run/dockershim.sock: connect: no such file or directory\"" image="europe-docker.pgk.dev/some-path/image:v0.1.0-alpha"
FATA[0000] pulling image: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial unix /var/run/dockershim.sock: connect: no such file or directory"
```

This suggests might also be related on how containerd has been setup. FYI: I performed a rather default installation of RKE2 (1.28) on Rocky9 via RPM packages. 

Could you please provide some assistance on how to circumnavigate this issue so that I can test the registries config? 

As mentioned RKE2 is currently working without apparente issues and is able to pull images by injecting the .dockerconfig as k8s secret. There must be something I'm missing here.

Thanks in advance.

**Reply by brandond** [reply]
It sounds like you've not set up your environment properly to talk to RKE2's containerd socket when using crictl? This is also covered in the docs.
https://docs.rke2.io/reference/cli_tools#containerd
```
export CRI_CONFIG_FILE=/var/lib/rancher/rke2/agent/etc/crictl.yaml
/var/lib/rancher/rke2/bin/crictl ps
```

**Reply by alex-ioma** [reply]
Thank you very much - definitely flew through that. Finally getting closer but still not working - error is: 
```bash
E0605 00:02:27.456638  357011 remote_image.go:171] "PullImage from image service failed" err="rpc error: code = Unknown desc = failed to pull and unpack image \"europe-docker.pkg.dev/repo/image:v0.1.0-alpha\": failed to resolve reference \"europe-docker.pkg.dev[...]": failed to authorize: failed to fetch oauth token: unexpected status from GET request to https://europe-docker.pkg.dev/v2/token?scope=repository%3A....: 401 Unauthorized
```

It seems that ctr is able to use the key provided through `registries.yaml` file, however, it is not able to use all of it's properties such as the fact that token should be obtain at the `token_uri`. 

For example: 
```json
{
  ...
  "auth_uri": "https://accounts.google.com/o/oauth2/auth",
  "token_uri": "https://oauth2.googleapis.com/token",
  "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
  "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/registry-ro%40cm-controls.iam.gserviceaccount.com",
  "universe_domain": "googleapis.com"
}
```

Are you aware of any issue on containerd side about this. Any suggestion on how to troubleshoot. 

Best regards.

**Reply by brandond** [reply]
I would probably go take a look at the upstream docs, and open an issue if necessary.

**Reply by alex-ioma** [reply]
I've tested different configurations also by interacting directly with config.toml.tmpl within containerd. Issue remains. 
I've open a ticket on the source. Will keep you posted.

**Reply by alex-ioma** [reply, marked as answer]
@brandond apologies for keeping this thread open - just realized that I did not circled back once the solution was found. 

Indeed it was as easy as adding these lines to the `registries.yaml` config: 

```yaml
configs:
  "europe-docker.pkg.dev":
    auth:
      username: _json_key_base64
      password: 'your base-64 formatted key file'
```
of course it also works with: 

```yaml
configs:
  "europe-docker.pkg.dev":
    auth:
      username: _json_key
      password: 'your key file here'
```

but in this latter case, you must pay attention to single apexes and special characters of the base-64 formatted certs that may break the `password` field. 

Also, there are sources/forums online that make use of `*.pkg.dev` as source. In our case, this has proven to be an issue as containerd or RKE2 expected strict URI and was not accepting `*` instead of `europe-docker`. If you still have issue, you may try to use the full URL.

Perhaps, incorporating these notes into the official docs might help others. 

Thanks for the great work.
